### ISSUE TITLE
[KV Feedback]: Better documentation of KV read/write units, what exactly is a write unit?

### ISSUE THREAD

**Issue by JoakimCh** [denoland/deploy_feedback#502]
> Free KV write units/day (1kb) 10,000

So what exactly is a write unit?
```js
kv.atomic()
  .set(key, value)
  .set(key, value)
.commit()
kv.atomic()
  .set(key, value)
  .set(key, value)
.commit()
```
Did the code use 4, 2 or 1?

**Comment by lucacasonato** [comment]
Depends on the size of value. `ceil((value size + key size of all mutations in the atomic operation) / 1024)`

So at least two, but up to 1600 (max size of a single atomic op is ~800kb)

**Comment by JoakimCh** [comment]
Okay. So if doing lots of writes it's best to batch them in an atomic then (to save cost)... Seems weird since initially atomic operations only supported 10 mutations because they were "heavy" (now 1000 is supported though).

Any failed check will ruin such a transaction though, hence it's not the ideal way to insert many if checks needs to be done.

**Comment by JoakimCh** [comment]
I think it would be better with a pricing system that more directly match our price impact on the underlying FoundationDB (if it could be measured). Because this feels very artificial.

EDIT: 
What about kv.getMany and kv.list? Will they also count as one unit if the data is below 4kb? A unit is basically each API call to the DB or what (and counted as multiple if over the KB limit)?

It would be nice to be able to calculate the size of the data e.g. a `kv.atomicDryRun(unitSize)` which returns `{numUnits, totalSize}`.

**Comment by cknight** [comment]
A few points.  I don't think you could get 1600 write units from 2 transactions as they are detailed above since both the key and value are capped at a maximum size (2kb and 64kb off the top of my head).  Secondly, it is FAR more efficient to write multiple items in one atomic transaction than individually.  If you have 100 items, this is one network request with an atomic transaction and 100 network requests as individual transactions.  

To answer your final question, my anecdotal viewing of my KV data is that kv.list streams the data and you are charged by the 4kb (for key and value size combined).  Size of key and value can roughly be calculated using JSON.stringify(value).length, though this is imperfect.  Thus if you have 10 key/values where each key/value is 300 bytes combined, then listing all 10 would be 1 read unit using kv.list.

**Comment by champkeh** [comment]
> I don't think you could get 1600 write units from 2 transactions as they are detailed above since both the key and value are capped at a maximum size (2kb and 64kb off the top of my head).

```ts
const _2K = Array.from({length: 1000 * 2}).fill('a').join('')
const _64K = Array.from({length: 1000 * 64}).fill('a').join('')
const _25K = Array.from({length: 1000 * 25}).fill('a').join('')

async function insertLargeData() {
    await kv.atomic()
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _64K)
        .set([_2K], _25K)
        .commit()
}
```
this 1 transactions will use 800 write units

**Comment by cknight** [comment]
> this 1 transactions will use 800 write units

That's correct, however I was commenting that a transaction with only 2 `set` operations (as shown in the OP example) couldn't hit the 1600 write units.
